_transcripts/050-rust-1.56-1.57.md

```diff
@@ -0,0 +1,1302 @@
+this question is entirely moot. The reason is, that on many OSes, Linux and
+macOS both, the OS just lies to you about how much memory you have. It is
+extremely common for the OS to say yes, I have memory, here you go, when in fact
+there is no memory backing it up. And it will kind of just like, lazily give you
+memory as you request it, or it'll do shenanigans with like, trying to compress
+memory if there is contention. Or they have too much stuff being given out. And
+so generally if the OS lies to you, there's really nothing you can do. And so
+many applications— the Rust standard library just doesn't, just doesn't care,
+really. And so it's like we— it would be an extremely burdensome API
+requirement, to require every single thing that could possibly fail to allocate
+memory, to bubble that up to the user. And so if it happens, then we'll just—
+we'll panic and— actually it'll abort, I believe, if that happens.
+
+But, if you are in an environment where you do— can guarantee that the OS won't
+lie to you, for example, if you're writing an OS then it's extremely useful to
+know that you are out of memory. And so this is kind of, it is not sufficient
+for Linux's use case, I believe. I haven't really read the proposal, but there
+was a long discussion. I'm not sure if it was a full RFC, but there have been
+many people involved with Linux, and contributing to the Linux kernel, have
+talked about how, okay, we would like to see various things from the standard
+library and Rust, to guarantee that we could use, even like, you know, libraries
+and (_unintelligible— 71:17_) I/O, and you know, guarantee that they're not
+going to allocate, you know, abort the process, you know, panic the kernel, just
+because they failed to allocate some memory. So this is kind of, maybe step one
+of that process. I'm sure there's much more to go, but so essentially, `Vec`,
+`String`, `HashMap`, `HashSet`, and `VecDeque` all have `try_reserve` as a new
+function.
+
+__Jon__: Yeah, and the big— one of the big missing parts here, I think, for the
+kernel, is that they want to sort of statically guarantee, so guarantee at
+compile time, that there's nothing in there that could panic because of out-of-
+memory, which— this is a step in the right direction, but it doesn't solve the
+problem, right? There is nothing that prevents someone from writing `.push` in
+the kernel, without first calling `try_reserve`. So there's still a question of,
+how can we statically guarantee the things that the Linux kernel requires us to
+guarantee. And that's still being figured out. But at least now we have the
+startings of the mechanisms that the kernel can use. And other use cases too,
+like embedded use cases, can use when they do want to know whether an allocation
+succeeded or failed.
+
+__Jon__: We have a bunch of other stabilized APIs too, which are sort of less,
+monumentus than `try_reserve`. I think the first one I wanted to call out is,
```

Comment:
I don't think "monumentus" is a word. Momentus? Not sure it this is worth correcting. (audio 72:40)

Reply:
Heh, yeah, it's an amalgamation of monumental and momentous, but not in a dictionary. I think it's fine to leave as-is.